src/main/scala/org/clulab/wm/eidos/context/TimeNormFinder.scala

```diff
@@ -93,34 +106,32 @@ class TimeNormFinder(parser: TemporalNeuralParser, timeRegexes: Seq[Regex]) exte
 
         // get the text for each context
         for (interval <- mergedIntervals) yield {
-          ((sentenceIndex, interval.start), sentenceText.substring(interval.start, interval.end))
+          (sentenceIndex,(interval.start, interval.end))
         }
       }
-
-      // TODO: TemporalNeuralParser should take Array arguments, not List arguments, so the .toList here can be removed
-      val (contextLocations: List[(Int, Int)], contextTexts: List[String]) = sentenceContexts.flatten.toList.unzip
+      val (contextSentenceIndexes: Array[Int], contextSpans: Array[(Int, Int)]) = sentenceContexts.flatten.unzip
 
       // run the timenorm parser over batches of contexts to find and normalize time expressions
-      val contextTimeExpressions: Seq[List[TimeExpression]] = eidosDocument.dctString match {
+      val contextTimeExpressions: Array[Array[TimeExpression]] = eidosDocument.dctString match {
         case Some(dctString) =>
-          val parsed = (dctString :: contextTexts).sliding(BATCH_SIZE, BATCH_SIZE).flatMap(parser.parse).toList
-          eidosDocument.dct = Some(DCT(parser.dct(parsed.head), dctString)) // Note the side effect!
-          parser.intervals(parsed.tail, eidosDocument.dct.map(_.interval))
+          val Array(dct: TimExInterval) = parser.parse(dctString)
+          // use a SimpleInterval (character span = None) so it doesn't interfere with character span of TimeExpressions
+          val dctSimpleInterval = SimpleInterval(dct.start, dct.end)
+          eidosDocument.dct = Some(DCT(dctSimpleInterval, dctString))
+          contextSpans.sliding(BATCH_SIZE, BATCH_SIZE).
+            flatMap(batch => parseBatch(text, batch, textCreationTime = eidosDocument.dct.get.interval)).toArray
         case None =>
-          val parsed = contextTexts.sliding(BATCH_SIZE, BATCH_SIZE).flatMap(parser.parse).toList
-          parser.intervals(parsed)
+          contextSpans.sliding(BATCH_SIZE, BATCH_SIZE).
+            flatMap(batch => parseBatch(text, batch)).toArray
       }
 
       // create mentions for each of the time expressions that were found
       for {
-        ((sentenceIndex, contextSentenceStart), timeExpressions) <- contextLocations zip contextTimeExpressions
+        (sentenceIndex, timeExpressions) <- contextSentenceIndexes zip contextTimeExpressions
         timeExpression <- timeExpressions
       } yield {
-        // reconstruct full-text character offsets from sentence-level character offsets
         val sentence = eidosDocument.sentences(sentenceIndex)
-        val contextStart = sentence.startOffsets.head + contextSentenceStart
-        val timeTextStart = contextStart + timeExpression.span.start
-        val timeTextEnd = contextStart + timeExpression.span.end
+        val (timeTextStart, timeTextEnd) = timeExpression.charSpan.get
```

Comment:
This is purely style, and not something you need to fix, but I prefer to write this as `val Some((timeTextStart, timeTextEnd)) = timeExpression.charSpan`